cli/src/main/java/hudson/cli/SSHCLI.java

```diff
@@ -0,0 +1,132 @@
+import static java.util.logging.Level.FINE;
+import java.util.logging.Logger;
+import org.apache.sshd.client.SshClient;
+import org.apache.sshd.client.channel.ClientChannel;
+import org.apache.sshd.client.channel.ClientChannelEvent;
+import org.apache.sshd.client.future.ConnectFuture;
+import org.apache.sshd.client.keyverifier.DefaultKnownHostsServerKeyVerifier;
+import org.apache.sshd.client.keyverifier.KnownHostsServerKeyVerifier;
+import org.apache.sshd.client.keyverifier.ServerKeyVerifier;
+import org.apache.sshd.client.session.ClientSession;
+import org.apache.sshd.common.future.WaitableFuture;
+import org.apache.sshd.common.util.SecurityUtils;
+import org.apache.sshd.common.util.io.NoCloseInputStream;
+import org.apache.sshd.common.util.io.NoCloseOutputStream;
+
+/**
+ * Implements SSH connection mode of {@link CLI}.
+ * In a separate class to avoid any class loading of {@code sshd-core} when not using {@code -ssh} mode.
+ * That allows the {@code test} module to pick up a version of {@code sshd-core} from the {@code sshd} module via {@code jenkins-war}
+ * that may not match the version being used from the {@code cli} module and may not be compatible.
+ */
+class SSHCLI {
+
+    static int sshConnection(String jenkinsUrl, String user, List<String> args, PrivateKeyProvider provider, final boolean strictHostKey) throws IOException {
+        Logger.getLogger(SecurityUtils.class.getName()).setLevel(Level.WARNING); // suppress: BouncyCastle not registered, using the default JCE provider
+        URL url = new URL(jenkinsUrl + "login");
+        URLConnection conn = url.openConnection();
+        String endpointDescription = conn.getHeaderField("X-SSH-Endpoint");
+
+        if (endpointDescription == null) {
+            CLI.LOGGER.warning("No header 'X-SSH-Endpoint' returned by Jenkins");
+            return -1;
+        }
+
+        CLI.LOGGER.log(FINE, "Connecting via SSH to: {0}", endpointDescription);
+
+        int sshPort = Integer.parseInt(endpointDescription.split(":")[1]);
+        String sshHost = endpointDescription.split(":")[0];
+
+        StringBuilder command = new StringBuilder();
+
+        for (String arg : args) {
+            command.append(QuotedStringTokenizer.quote(arg));
+            command.append(' ');
+        }
+
+        try(SshClient client = SshClient.setUpDefaultClient()) {
+
+            KnownHostsServerKeyVerifier verifier = new DefaultKnownHostsServerKeyVerifier(new ServerKeyVerifier() {
+                @Override
+                public boolean verifyServerKey(ClientSession clientSession, SocketAddress remoteAddress, PublicKey serverKey) {
+                    CLI.LOGGER.log(Level.WARNING, "Unknown host key for {0}", remoteAddress.toString());
+                    return !strictHostKey;
+                }
+            }, true);
+
+            client.setServerKeyVerifier(verifier);
+            client.start();
+
+            ConnectFuture cf = client.connect(user, sshHost, sshPort);
+            cf.await();
```

Comment:
> in which case the client waits and the user eventually kills it

Jenkins CLI is also a tool for automation. Such possible hanging is not a good UX IMHO. Can convert to Ant if you feel strongly && can handle the possible fallout later if it happens